network/crl_test.go

```diff
@@ -0,0 +1,392 @@
+uzvBbX+4BtopierZdEXysa3WoFpM7rIxryOYeXegDjiDFqY7RmZvs9cD9hB+SDaX
+W7whTRh8jFWOmGVEAq4YbiqbkfaFEX+8YrFoAjl4zHFlXfE3K7Wvx/WxGYuS50+4
+4Wofkd6afnVggqWli3gLn/iYLjbog94vYJ0CSBf13NfYn8frMa01O9GGHsuShjNJ
+MDt+/ZfPoHPVmQIDAQABo4GUMIGRMAkGA1UdEwQCMAAwCwYDVR0PBAQDAgXgMBQG
+A1UdEQQNMAuCCWxvY2FsaG9zdDAxBggrBgEFBQcBAQQlMCMwIQYIKwYBBQUHMAGG
+FWh0dHA6Ly8xMjcuMC4wLjE6ODg4ODAuBgNVHR8EJzAlMCOgIaAfhh1odHRwOi8v
+MTI3LjAuMC4xOjg4ODkvY3JsLnBlbTANBgkqhkiG9w0BAQsFAAOCAQEAH4yh4kUT
+wAxm7cKNu73hOGwz9HFswp44jpAl4q9/kJHkA/ycITFbHEPKsruWfqLk7vxp7kgW
+8Fv+qMyP5b0njDtAUWQuV8gn+DRQzPun8scG0kGJQb24ZQ4nWEPFm42H/roYv47B
+7y7cYmnycP9c4oCaQMdJ29w4bdaC9SHZdQW0W57UkmelVjFU00odxia5O7VhKRkP
+FC+mY9/+xkj2aDvunWLHwE4wXSzrFpOkGqNPpRU47HNO3K3giRTZXtKCap61J55B
+W8Q1chrNalbgBuYfsbtVlCEoVuzPmkjGzJIHASk8NNo3jQ7EgQJUbff8d1MZwa2f
+k5rkNiHTxPfguQ==
+-----END CERTIFICATE-----
+`)
+}
+
+func getTestCRL() []byte {
+	// x509.ParseCRL() that is called inside CRL verifier is very fragile
+	// and will return error because of \n on start, that's why this func
+	// looks a little bit different from getTest*Cert()
+
+	// tests/ssl/crl.pem
+	return []byte(`-----BEGIN X509 CRL-----
+MIIB4zCBzDANBgkqhkiG9w0BAQsFADB0MQswCQYDVQQGEwJHQjEPMA0GA1UECAwG
+TG9uZG9uMQ8wDQYDVQQHDAZMb25kb24xGDAWBgNVBAoMD0dsb2JhbCBTZWN1cml0
+eTELMAkGA1UECwwCSVQxHDAaBgNVBAMME1Rlc3QgQ0EgY2VydGlmaWNhdGUXDTIw
+MTAyNjIzMjcwNloXDTIwMTEyNTIzMjcwNlowJzAlAhQH6gdHXIXsVUmzREIYCF6n
+ajAdThcNMjAxMDI2MjMyNzA2WjANBgkqhkiG9w0BAQsFAAOCAQEAKKJ56Ba7bOqE
+R8LFhfSwqqzGxVdPfDOu3d14cpRC3PJSLcyfLh+YA8Rkveed7Ee1RGaGjXrXTiZm
+p44CQcslUFoFGdONVTzTAamCphOiLh/LZB65d2dWo1WwOfsLwwaE+KTk+99+Ug/j
+EgiyuOwoyFqONlaDj4Z2PUC/mKmPuI1QJ9WB43oL8cDxPuOtJp0z2NEz0ZDl1pu6
+ghaEHc7lPDBmCOOGYneqEtfAi/wzMzKEdeXxV9s3MBWsXKfdrwsQc2pjLuH+K/wv
+aY3BPTUA7sZLJGFvsE2fh9yi1TsyKs4hsboS6FldZdXBnD/0ZqUnFy/KQ4t3LwH5
+L6zfD+wegQ==
+-----END X509 CRL-----`)
+}
+
+// Convert PEM-encoded certificate into DER and parse it into x509.Certificate
+func pemToX509Cert(data []byte) (*x509.Certificate, error) {
+	certDERBlock, _ := pem.Decode(data)
+	if certDERBlock == nil {
+		return nil, errors.New("Failed to decode PEM")
+	}
+
+	if certDERBlock.Type != "CERTIFICATE" {
+		return nil, errors.New("Decoded DER block was not a certificate")
+	}
+
+	cert, err := x509.ParseCertificate(certDERBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+func TestDefaultCRLClientHTTP(t *testing.T) {
+	crl := getTestCRL()
+	addr := "127.0.0.1:45318"
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/test_crl.pem", func(res http.ResponseWriter, req *http.Request) {
+		res.WriteHeader(200)
+		res.Header().Add("Content-Type", "application/pem-certificate-chain")
+		res.Write(crl)
+	})
+
+	httpServer := &http.Server{
+		Addr:    addr,
+		Handler: mux,
+	}
+	go func() {
+		httpServer.ListenAndServe()
+	}()
+	defer httpServer.Close()
+
+	// Give the server time to start, otherwise this test will fail sometimes
+	time.Sleep(time.Millisecond * 50)
```

Comment:
better to add one more function (like check_connection in our python tests) which will test HTTP server's startups in a loop with some retry count and sleeps, after which return control or call t.Fatal. 
We had situations when on local machine was enough 50ms to start some service but had blinking tests in CI environments where tests running on slower machines and in docker containers

Reply:
Okay, will add some kind of loop. However, those problems you are talking about, they are more related to multi-container or at least multi-process setups while here we have everything in a single process.